src/lib/optimizer/strategy/data_induced_predicate_rule.cpp

```diff
@@ -0,0 +1,130 @@
+#include "data_induced_predicate_rule.hpp"
+
+#include "cost_estimation/abstract_cost_estimator.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/enable_make_for_lqp_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "statistics/abstract_cardinality_estimator.hpp"
+
+namespace hyrise {
+using namespace expression_functional;  // NOLINT(build/namespaces)
+
+std::string DataInducedPredicateRule::name() const {
+  static const auto name = std::string{"DataInducedParameterRule"};
+  return name;
+}
+
+void DataInducedPredicateRule::_apply_to_plan_without_subqueries(
+    const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
+  Assert(lqp_root->type == LQPNodeType::Root, "Rule needs root to hold onto");
+
+  // Adding dips inside visit_lqp might lead to endless recursions. Thus, we use visit_lqp to identify the
```

Comment:
It's not really about recursions. First, visit_lqp is not recursive but a breadth-first search w/ node caching (i.e., exactly one visit). Second, it only visits outputs of the **original** input – added or removed nodes are not considered. The main arguments not do directly insert the nodes are (i) that we do not want to consider the interrelations between multiple diPs (especially if we reduce both sides) and (ii) that we guarantee the cardinality estimator not to modify the LQP for/below already estimated nodes to make use of statistics caching (`guarantee_bottom_up_construction()`).